New banhdid and ban

Poyoanon opened the pull request:
Proper logging of them as well.

Poyoanon committed:
targeted id check for mods

more secrets

Poyoanon committed:
Merge pull request #3 from ghostfeesh/targeted-id-check

targeted id check for mods

Poyoanon committed:
some /ban changes

fixed a bug with /kick or something
added something new to /ban

Poyoanon committed:
Merge pull request #4 from ghostfeesh/ban-patch

some /ban changes

collinxchu committed:
Update client_manager.py

collinxchu committed:
Update client_manager.py

collinxchu committed:
Update hdidbans to actually work.

collinxchu committed:
Add /unbanhdid, small change to /unban

/unbanhdid will remove an hdidban, and output a log as well as an OOC message.
/unban will now output an OOC message upon success.

Poyoanon committed:
Merge pull request #6 from collinxchu/master

Now HDID bans and IP bans work, Plus logging.

collinxchu committed:
Update README.md

Poyoanon committed:
Merge pull request #7 from collinxchu/patch-1

Update README.md

oldmud0 commented on server/ban_manager.py:
```diff
@@ -32,10 +34,21 @@ def load_banlist(self):
+    def load_hdidbanlist(self):
+        try:
+            with open('storage/hdidbanlist.json', 'r') as hdidbanlist_file:
+                self.hdidbans = json.load(hdidbanlist_file)
```
you need to handle `ValueError` as well if you don't want it to crash on JSON parse error

collinxchu replied:
Will look into it.

oldmud0 commented on server/client_manager.py:
```diff
@@ -288,7 +288,14 @@ def get_targets_by_ip(self, ip):
+        for client in self.clients:
+            if client.get_hdid() == hdid:
+                clients.append(client)
+        return clients
```
you can use a generator/`yield` here but I get that you want to be consistent with the code.

collinxchu replied:
To be honest I wouldn't have the foggiest clue what any of that means.

oldmud0 commented on server/client_manager.py:
```diff
@@ -324,4 +331,28 @@ def get_ooc_muted_clients(self):
+            for client in self.clients:
+                if client.get_hdid() is not None and client.get_ip() == ip:
+                    hdid = client.get_hdid()
+                    return hdid
```
this would probably be less confusing if you used your new-fangled `get_targets_by_hdid`. actually nope I'm not understanding why this has to be O(n^2)

collinxchu replied:
Essentially it's attempting to compare based on current connections to see if the person you're banning has an hdid for logging purposes.

hdids aren't stored in an easily-accessible location, so fetching from the current list of connected users in the area is the best I could come up with.

oldmud0 replied:
I wonder why you can't do:
```python
for client in self.clients:
    if client.get_ip() == ip:
        return client.get_hdid()
```

collinxchu replied:
Fixed in commits 801b908 and 5565efa

oldmud0 commented on server/client_manager.py:
```diff
@@ -324,4 +331,28 @@ def get_ooc_muted_clients(self):
+        if hdid == None:
+            return None
+        else:
+            return hdid
```
you can condense to `return hdid or None`

collinxchu replied:
I can't confess to knowing exactly what that would look like.

`if hdid == None:`
`    return hdid or None`

???

oldmud0 commented on server/client_manager.py:
```diff
@@ -324,4 +331,28 @@ def get_ooc_muted_clients(self):
+        if ip == None:
+            return None
+        else:
+            return ip
```
ditto

oldmud0 commented on server/commands.py:
```diff
@@ -617,3 +631,55 @@ def ooc_cmd_mutepm(client, arg):
+    if not client.is_mod:
+        raise ClientError('You must be authorized to do that.')
+    ip = arg.strip()
+    if len(ip) < 7:
```
Where is this magical 7 from? If you want to check for a valid IP address, you should at least do it right. IPv6 addresses are quite tricky.

collinxchu replied:
Probably ought to change that - it was in the original function, however.

oldmud0 commented on server/commands.py:
```diff
@@ -617,3 +631,55 @@ def ooc_cmd_mutepm(client, arg):
+    targets = client.server.client_manager.get_targets_by_ip(ip)
+    if targets:
+        for c in targets:
+            c.disconnect()
```
No notice of kick/ban?

collinxchu replied:
The notice is that you are disconnected. I have no idea how one could display the message that you've been kick/banned within the AO client framework anyway.

oldmud0 commented on server/commands.py:
```diff
@@ -425,6 +425,20 @@ def ooc_cmd_unban(client, arg):
+    if not client.is_mod:
+        raise ClientError('You must be authorized to do that.')
+    hdid = arg.strip()
+    if len(hdid) < 8:
```
Since when did HDIDs have to be at least 8 characters long?

collinxchu replied:
That's how it was in the original code. I have not changed anything here. I don't actually know shit about how hdid works.

oldmud0 replied:
If you prevent short HDIDs from being parsed, you are essentially disallowing yourself from banning potentially abusive and invalid HDIDs.

collinxchu replied:
fixed in commit 801b908

oldmud0 commented on server/commands.py:
```diff
@@ -617,3 +631,55 @@ def ooc_cmd_mutepm(client, arg):
+    targets = client.server.client_manager.get_targets_by_hdid(hdid)
+    if targets:
+        for c in targets:
+            c.disconnect()
```
You should notify users of their kick/ban

collinxchu replied:
I don't really know an elegant way of doing that, sadly. What with a disconnect already creating a popup client-side.

Poyoanon committed:
snek 3.6.2

Poyoanon committed:
Update README.md

Poyoanon committed:
Update README.md

collinxchu committed:
Sanity Checks

Now the hdid and ip bans make sure that the variables they want exist.

And tell you if you've done something wrong.

Poyoanon committed:
Merge pull request #9 from collinxchu/master

Sanity Checks

collinxchu committed:
Account for ValueError exception

oldmud0 commented on server/client_manager.py:
```diff
@@ -288,7 +288,14 @@ def get_targets_by_ip(self, ip):
+        for client in self.clients:
+            if client.get_hdid() == hdid:
+                clients.append(client)
+        return clients
```
Instead of declaring a list ahead of time, appending to it, and returning the list, you can just replace the line in the inner for loop that adds the client to the list to say `yield client`. Then you do not need a list or a return statement.

oldmud0 commented on server/client_manager.py:
```diff
@@ -324,4 +331,28 @@ def get_ooc_muted_clients(self):
+        if hdid == None:
+            return None
+        else:
+            return hdid
```
You can remove the whole if/else block and replace it with `return hdid or None`

oldmud0 commented on server/commands.py:
```diff
@@ -395,24 +394,66 @@ def ooc_cmd_kick(client, arg):
-    if len(ip) < 7:
-        raise ArgumentError('You must specify an IP.')
+    chk = ip.replace(".", "")
+    if len(ip) < 7 or not chk.isdigit():
```
This is not IPv6-compatible. Use the [`ipaddress`](https://docs.python.org/3/library/ipaddress.html) module and handle `ValueError` on error.

collinxchu replied:
Fixed in commit 0a460e6

oldmud0 commented on server/commands.py:
```diff
@@ -395,24 +394,66 @@ def ooc_cmd_kick(client, arg):
+        client.send_host_message('Kicked {} existing client(s).'.format(len(targets)))
+    if ip != None:
+        client.send_host_message('Added {} to the banlist.'.format(hdid))
+        logger.log_server('Banned {}, {} at HDID.'.format(ip, hdid), client)
```
`by HDID`*

collinxchu replied:
Fixed in commit 5a7e654

oldmud0 commented on server/commands.py:
```diff
@@ -395,24 +394,66 @@ def ooc_cmd_kick(client, arg):
+        logger.log_server('Banned {}, {} at HDID.'.format(ip, hdid), client)
+    else:
+        client.send_host_message('Added {} to the banlist.'.format(hdid))
+        logger.log_server('Banned {} at HDID. Could not locate client ip.'.format(hdid), client)
```
ditto

collinxchu replied:
Fixed in commit 5a7e654
